src/components/lib/shm_bm/shm_bm.c

```diff
@@ -0,0 +1,226 @@
+}
+
+/**
+ * Round input to the next power of 2. From https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+ * but generalized for a generic word size.
+ */
+static word_t
+shm_bm_npow2(word_t w)
+{
+	unsigned long s;
+
+	w--;
+	for (s = 1; s < (sizeof (word_t) * CHAR_BIT); s *= 2) {
+		w |= (w >> s);
+	}
+
+  	return ++w;
+}
+
+static inline int
+shm_bm_log2(word_t x)
+{
+	if (x == 0) return 0;
+    return (sizeof (word_t) * CHAR_BIT) - __builtin_clzl(x) - 1;
+}
+
+cbuf_t   
+shm_bm_create(shm_bm_t *shm, size_t objsz, int nobj)
+{
+	struct shm_bm_header *header;
+	size_t                bitmap_sz, refcnt_sz, data_sz;
+	size_t                objszp2, alloc;
+	cbuf_t                id;
+
+	objszp2 = shm_bm_npow2(objsz);
+	if (nobj <= 0) return 0;
+
+	bitmap_sz = SHM_BM_BITS_TO_WORDS(nobj) * sizeof (word_t);
+	refcnt_sz = nobj;
+	data_sz   = nobj * objszp2;
+
+	alloc = sizeof (struct shm_bm_header) + bitmap_sz + refcnt_sz + data_sz;
+	if (alloc > SHM_BM_ALLOC_ALIGNMENT) return 0;
+
+	id  = memmgr_shared_page_allocn_aligned(round_up_to_page(alloc)/PAGE_SIZE, SHM_BM_ALLOC_ALIGNMENT, (vaddr_t *) shm);
+	
+	if (id == 0) return 0;
+	memset((void *) *shm, 0, round_up_to_page(alloc));
+
+	// metadata
+	header = (struct shm_bm_header *) *shm;
+	header->objsz_order = shm_bm_log2(objszp2);
+	header->nobj        = nobj;
+	header->refc_offset = sizeof (struct shm_bm_header) + bitmap_sz; 
+	header->data_offset = sizeof (struct shm_bm_header) + bitmap_sz + refcnt_sz; 
+
+	// set nobj bits to free
+	shm_bm_set_contig(SHM_BM_BITM(*shm), nobj);
+	return id;
+}
+
+shm_bm_t 
+shm_bm_map(cbuf_t id)
+{
+	shm_bm_t shm;
+
+	if (memmgr_shared_page_map_aligned(id, SHM_BM_ALLOC_ALIGNMENT, (vaddr_t *) &shm) == 0) return 0;
+
+	return shm;
+}
+
+void *
+shm_bm_obj_alloc(shm_bm_t shm, shm_bufid_t *id)
+{
+	int     freebit, idx , offset, lz;
+	word_t  word; 
+	word_t *bm;
+
+	/* 
+	 * Find a free space; could be preempted. Find the next word in
+	 * the bitmap with a free bit and get its index in the bitmap, and 
+	 * the offset of the free bit in that word. Then try to atomically 
+	 * xchg the bitmap with the freebit set to 0
+	 */
+	bm = SHM_BM_BITM(shm);
+	do {
+		idx = shm_bm_next_free_word(bm, SHM_BM_REFC(shm));
+		if (idx == -1) return 0;
+		word   = bm[idx];
+		lz     = __builtin_clzl(word); 
+		offset = SHM_BM_BITMAP_BLOCK - lz - 1;
+	} while (!cos_cas(bm + idx, word, word & ~(1ul << offset)));
+
+	freebit = lz + (idx * SHM_BM_BITMAP_BLOCK);
+	cos_faab(SHM_BM_REFC(shm) + freebit, 1);
+
+	*id = (shm_bufid_t) freebit;
+	return SHM_BM_DATA(shm) + (freebit << SHM_BM_OBJSZ_ORDER(shm));
+}
+
+void *   
+shm_bm_obj_use(shm_bm_t shm, shm_bufid_t id)
+{
+	if (unlikely(id >= SHM_BM_NOBJ(shm))) return 0;
+
+	// obj has not been allocated
+	if (unlikely(*(SHM_BM_REFC(shm) + id) == 0)) return 0;
+
+	cos_faab(SHM_BM_REFC(shm) + id, 1);
+
+	return SHM_BM_DATA(shm) + (id << SHM_BM_OBJSZ_ORDER(shm));
+}
+
+void *   
+shm_bm_obj_take(shm_bm_t shm, shm_bufid_t id)
```

Comment:
"Take" usually means "refcnt increase" or "take lock". I think that you might have reversed "take" and "use"?